src/connection-manager/dial-queue.ts

```diff
@@ -0,0 +1,498 @@
+        // remove our pending dial entry
+        this.pendingDials = this.pendingDials.filter(p => p.id !== pendingDial.id)
+
+        // clear the dial timeout if it's not fired already
+        timeoutController.clear()
+      })
+      .catch(err => {
+        log.error('dial failed to %s', addrsToDial.map(({ multiaddr }) => multiaddr.toString()).join(', '), err)
+
+        // Error is a timeout
+        if (timeoutController.signal.aborted) {
+          const error = new CodeError(err.message, codes.ERR_TIMEOUT)
+          throw error
+        }
+
+        throw err
+      })
+
+    // let other dials join this one
+    this.pendingDials.push(pendingDial)
+
+    return await pendingDial.promise
+  }
+
+  private createDialAbortControllers (userSignal?: AbortSignal): { timeoutController: TimeoutController, signal: AbortSignal } {
+    // ensure we throw if the dial takes longer than the dial timeout
+    const timeoutController = new TimeoutController(this.dialTimeout)
+
+    // let any signal abort the dial
+    const signal = combineSignals(
+      timeoutController.signal,
+      this.shutDownController.signal,
+      userSignal
+    )
+
+    return { timeoutController, signal }
+  }
+
+  private async calculateMultiaddrs (peerId?: PeerId, addrs: Address[] = [], options: DialOptions = {}): Promise<Address[]> {
+    // if a peer id or multiaddr(s) with a peer id, make sure it isn't our peer id and that we are allowed to dial it
+    if (peerId != null) {
+      if (this.peerId.equals(peerId)) {
+        throw new CodeError('Tried to dial self', codes.ERR_DIALED_SELF)
+      }
+
+      if ((await this.connectionGater.denyDialPeer?.(peerId)) === true) {
+        throw new CodeError('The dial request is blocked by gater.allowDialPeer', codes.ERR_PEER_DIAL_INTERCEPTED)
+      }
+
+      // if just a peer id was passed, load available multiaddrs for this peer from the address book
+      if (addrs.length === 0) {
+        log('loading multiaddrs for %p', peerId)
+        addrs.push(...(await this.peerStore.addressBook.get(peerId)))
+        log('loaded multiaddrs for %p', peerId, addrs.map(({ multiaddr }) => multiaddr.toString()))
+      }
+    }
+
+    // resolve addresses - this can result in a one-to-many translation when dnsaddrs are resolved
+    const resolvedAddresses = (await Promise.all(
+      addrs.map(async addr => {
+        const result = await resolveMultiaddrs(addr.multiaddr, options)
+
+        if (result.length === 1 && result[0].equals(addr.multiaddr)) {
+          return addr
+        }
+
+        return result.map(multiaddr => ({
+          multiaddr,
+          isCertified: false
+        }))
+      })
+    ))
+      .flat()
+
+    // filter out any multiaddrs that we do not have transports for
+    const filteredAddrs = resolvedAddresses.filter(addr => Boolean(this.transportManager.transportForMultiaddr(addr.multiaddr)))
+
+    // deduplicate addresses
+    const dedupedAddrs: Map<string, Address> = new Map()
```

Comment:
Couldn't we do something like 

```suggestion
let dedupedMultiaddrs = [...new Set(filteredAddrs.map(addr => { 
	addr.isCertified = addr.isCertified || addr.isCertified || false
})).values()]
```

to avoid the null checks per item

Reply:
`Address`es are not `Multiaddr`s - they have an `isCertified` field so we can't just stringify them into a set to dedupe, instead we use the keys of a map to dedupe but using the `Address` as map values lets us preserve the `isCertified` field.